src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java

```diff
@@ -0,0 +1,292 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nutch.protocol.smb;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.apache.nutch.crawl.CrawlDatum;
+import org.apache.nutch.metadata.Metadata;
+import org.apache.nutch.net.protocols.HttpDateFormat;
+import org.apache.nutch.protocol.Content;
+import org.apache.nutch.protocol.Protocol;
+import org.apache.nutch.protocol.ProtocolOutput;
+import org.apache.nutch.protocol.ProtocolStatus;
+import org.apache.nutch.protocol.RobotRulesParser;
+import com.hierynomus.msdtyp.AccessMask;
+import com.hierynomus.msfscc.FileAttributes;
+import com.hierynomus.msfscc.fileinformation.FileAllInformation;
+import com.hierynomus.msfscc.fileinformation.FileIdBothDirectoryInformation;
+import com.hierynomus.mssmb2.SMB2CreateDisposition;
+import com.hierynomus.mssmb2.SMB2CreateOptions;
+import com.hierynomus.mssmb2.SMB2ShareAccess;
+import com.hierynomus.smbj.auth.AuthenticationContext;
+import com.hierynomus.smbj.connection.Connection;
+import com.hierynomus.smbj.session.Session;
+import com.hierynomus.smbj.share.DiskShare;
+import com.hierynomus.smbj.share.File;
+import com.hierynomus.smbj.SMBClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import crawlercommons.robots.BaseRobotRules;
+
+public class Smb implements Protocol {
+  protected static final Logger LOG = LoggerFactory.getLogger(Smb.class);
+
+  private Configuration conf;
+
+  private String user;
+  private String password;
+  private String domain;
+  private int contentLimit;
+  private Set<String> ignoreFiles;
+
+  public Smb() {
+    // todo: files that should be skipped could be configurable.
+    this.ignoreFiles = new HashSet<>();
+    ignoreFiles.add(".");
+    ignoreFiles.add("..");
+    ignoreFiles.add(".svn");
+    ignoreFiles.add(".git");
+  }
+
+  @Override
+  public Configuration getConf() {
+    LOG.debug("getConf()");
+    return this.conf;
+  }
+
+  @Override
+  public void setConf(Configuration conf) {
+    this.conf = conf;
+
+    // todo: is it possible to use configuration "per server" or "per share"?

```

Comment:
You can take a look at [IndexWriters#loadWritersConfiguration(Configuration conf)](https://github.com/apache/nutch/blob/4a61208f492613f2c5282741e64c036acabeb71e/src/java/org/apache/nutch/indexer/IndexWriters.java#L116-L143). Here you can see that we [read an XML file](https://github.com/apache/nutch/blob/4a61208f492613f2c5282741e64c036acabeb71e/conf/index-writers.xml.template) containing several indexing writer plugin configurations.
Maybe we could do something similar for the SMB plugin. This would set the standard for how other protocol plugins could accomplish the same e.g. SMTP, (S)FTP, etc.